crates/binstalk-downloader/src/gh_api_client/request.rs

```diff
@@ -0,0 +1,133 @@
+use std::{collections::HashSet, io, time::Duration};
+
+use compact_str::CompactString;
+use serde::Deserialize;
+use serde_json::Deserializer as JsonDeserializer;
+use thiserror::Error as ThisError;
+use url::Url;
+
+pub use serde_json::Error as JsonError;
+
+use super::{remote, GhRelease};
+use crate::utils::{extract_with_blocking_task, StreamReadable};
+
+#[derive(ThisError, Debug)]
+pub enum GhApiError {
+    #[error("IO Error: {0}")]
+    Io(#[source] io::Error),
+
+    #[error("Failed to parse json: {0}")]
+    Json(#[from] JsonError),
+
+    #[error("Remote Error: {0}")]
+    Remote(#[from] remote::Error),
+
+    #[error("Failed to parse url: {0}")]
+    InvalidUrl(#[from] url::ParseError),
+}
+
+impl From<io::Error> for GhApiError {
+    fn from(err: io::Error) -> Self {
+        match err.get_ref() {
+            Some(inner_err) if inner_err.is::<Self>() => {
+                let inner = err.into_inner().expect(
+                    "err.get_ref() returns Some, so err.into_inner() should also return Some",
+                );
+
+                inner.downcast().map(|b| *b).unwrap()
+            }
+            Some(inner_err) if inner_err.is::<JsonError>() => {
+                let inner = err.into_inner().expect(
+                    "err.get_ref() returns Some, so err.into_inner() should also return Some",
+                );
+
+                inner.downcast().map(|b| GhApiError::Json(*b)).unwrap()
+            }
+            _ => GhApiError::Io(err),
+        }
+    }
+}
+
+// Only include fields we do care about
+
+#[derive(Deserialize)]
+struct Asset {
+    name: CompactString,
+}
+
+#[derive(Deserialize)]
+struct Response {
+    assets: Vec<Asset>,
+}
+
+pub enum FetchReleaseRet {
+    ReachedRateLimit { retry_after: Option<Duration> },
+    ReleaseNotFound,
+    Artifacts(HashSet<CompactString>),
+}
+
+/// Returns 404 if not found
+pub(super) async fn fetch_release_artifacts(
+    client: &remote::Client,
+    GhRelease { owner, repo, tag }: GhRelease,
+    auth_token: Option<&str>,
+) -> Result<FetchReleaseRet, GhApiError> {
+    let mut request_builder = client
```

Comment:
can we do a graphql query instead? likely to be a lot more efficient esp api limits wise

Reply:
Not very familiar with graphql, I will look into it.

Reply:
I would have to first fix the bugs in this PR first.

Reply:
Something like this

```graphql
query { 
  cargoBinstall: repository(owner: "cargo-bins", name: "cargo-binstall") {
    withV: release(tagName: "v0.20.1") { ...artifacts }
    noV: release(tagName: "0.20.1") { ...artifacts }
  }

  cargoWatch: repository(owner: "watchexec", name: "cargo-watch") {
    withV: release(tagName: "v8.4.0") { ...artifacts }
    noV: release(tagName: "8.4.0") { ...artifacts }
  }
}

fragment artifacts on Release {
  isLatest
  isPrerelease
  releaseAssets(first: 100) {
    nodes {
      name
      url
    }
  }
}
```

Reply:
I think we'd still need the ReST API for unauthenticated acccess, tho

Reply:
> I think we'd still need the ReST API for unauthenticated acccess, tho

Uhhh then it perhaps isn't worth the time supporting this GraphQL for now.

> would do a single request for all the lookups at once.

That will be great and reduce the network overhead, but requires some architecture changes to fetch everything at once.

Considering that GraphQL is only available with a token, I'm a bit hesitant to do this.

Reply:
> likely to be a lot more efficient esp api limits wise

Since GraphQL cannot be used for unauthorized access and for access with token, GitHub allows 5000 requests per hous, I don't think we can adopt GraphQL api to avoid hitting rate limit.

Besides, it seems that GitHub has its own way of calculating rate limit for GraphQL since it is very versatile and can incur a lot of database query in one request, so we might just keep using and improving our use of the existing GitHub Restful API.

Reply:
> Besides, it seems that GitHub has its own way of calculating rate limit for GraphQL

oh I tested that and we can do something like 5k crates per hour with the quota in there.

But agreed on other points.

Reply:
> oh I tested that and we can do something like 50k crates per hour with the quota in there.

50k crates per hour?
Using the GitHub Restful API we can only check 5k crates per hous, 50k is a lot.

In that case, maybe @passcod you can submit another PR after this one to add a specialisation if the GITHUB_TOKEN is provided?

That might be useful to further speedup resolution and for repositories that might hit the limit, though it's hard to me to imagine hitting the 5k limit already.